components/push/ffi/src/lib.rs

```diff
@@ -0,0 +1,207 @@
+        let options = conn.options.clone();
+        let channel = ffi_support::rust_str_from_c(channel_id);
+        let key = options.vapid_key;
+        let reg_token = options.registration_id;
+        let subscription_key = crypto::generate_key().unwrap();
+        let auth = base64::encode_config(
+            &crypto::get_bytes(SER_AUTH_LENGTH), base64::URL_SAFE_NO_PAD);
+        // Don't auto add the subscription to the db.
+        // (endpoint updates also call subscribe and should be lighter weight)
+        let info = conn.subscribe(channel).unwrap();
+        // store the channelid => auth + subscription_key
+        let mut record = PushRecord::new(
+            &info.uaid,
+            &channel,
+            &info.endpoint,
+            "",
+            subscription_key.clone());
+        record.app_server_key = options.vapid_key.clone();
+        record.native_id = Some(reg_token);
+        self.database.put_record(&record)?;
+        let subscription_info = json!({
+            "endpoint": info.endpoint
+            "keys": {
+                "auth": auth,
+                "p256dh": base64::encode_config(subscription_key.public,
+                                                base64::URL_SAFE_NO_PAD)
+            }
+        });
+        return subscription_info.to_string()
+    })
+}
+
+// Unsubscribe a channel
+#[no_mangle]
+pub unsafe extern "C" fn push_unsubscribe(
+    handle: u64,
+    channel_id: *const c_char,
+    error: &mut ExternError,
+) -> u8 {
+    log::debug!("push_unsubscribe");
+    CONNECTIONS.call_with_result_mut(error, handle, |conn| {
+        let channel = ffi_support::opt_rust_str_from_c(channel_id);
+        conn.subscribe(channel).unwrap()
+    })
+}
+
+// Update the OS token
+#[no_mangle]
+pub unsafe extern "C" fn push_update(
+    handle: u64,
+    new_token: *const c_char,
+    error: &mut ExternError,
+) -> u8 {
+    log::debug!("push_update");
+    CONNECTIONS.call_with_result_mut(error, handle, |conn| {
+        let token = ffi_support::opt_rust_str_from_c(new_token);
+        conn.update(token).unwrap()
+    })
+}
+
+// verify connection using channel list
+// Returns a JSON containing the new channelids => endpoints
+// NOTE: AC should notify processes associated with channelIDs of new endpoint
+#[no_mangle]
+pub unsafe extern "C" fn push_verify_connection(
+    handle: u64,
+    error: &mut ExternError,
+) -> *mut c_char{
+    log::debug!("push_verify");
+    CONNECTIONS.call_with_result_mut(error, handle, |conn| {
+        let known_channels = storage::get_channel_list(conn.uaid.unwrap());
+        let options = conn.options.clone();
+        let key = options.vapid_key;
+        let reg_id = options.registration_id;
+        if ! conn.verify_connection(known_channels){
+            if let Some(new_endpoints) = conn.regenerate_endpoints(
+                known_channels,
+                key,
+                reg_id
+                ){
+                    serde_json::to_string(new_endpoints)
```

Comment:
FYI if you [`implement_into_ffi_by_json!`](https://docs.rs/ffi-support/0.2.0/ffi_support/macro.implement_into_ffi_by_json.html) you don't need to so the to_string manually.

This implements `IntoFfi` for `T` (by converting to JSON and returning `*mut c_char`), `Vec<T>` (same way), and `Option<T>` (also same way, but returns `std::ptr::null_mut()` for `None`).